File: ios/NativeLogsModule.swift
import Foundation
import os.log
import OSLog
import React

@objc(NativeLogsModule)
class NativeLogsModule: RCTEventEmitter {

  private var logging = false
  private var logThread: Thread?

  override static func requiresMainQueueSetup() -> Bool {
      return true
  }
  
  @available(iOS 15.0, *)
  @objc(startLogging)
  func startLogging() {
      if logging { return }
      logging = true

      self.sendLog("Logging started")
      logThread = Thread {
          self.fetchLogs()
      }
      logThread?.start()
  }
  
  @objc(stopLogging)
     func stopLogging() {
         self.sendLog("Logging stopped")
         if !logging { return }
         logging = false
         logThread?.cancel()
         logThread = nil
     }

  
  @available(iOS 15.0, *)
     private func fetchLogs() {
         guard let logStore = try? OSLogStore(scope: .currentProcessIdentifier) else { return }
         let position = logStore.position(timeIntervalSinceLatestBoot: 0)
         let logBuffer: StringBuffer = StringBuffer()

         while logging {
             do {
                 let entries = try logStore.getEntries(at: position)
                 for entry in entries {
                     if let logEntry = entry as? OSLogEntryLog {
                       if let logEntry = entry as? OSLogEntryLog, logEntry.category == "", logEntry.subsystem == "" {
                         logBuffer.append(logEntry.composedMessage + "\n")
                       }
                     }
                 }
             } catch {
                 print("Error fetching logs: \(error)")
             }

             Thread.sleep(forTimeInterval: 1)
             if logBuffer.isNotEmpty() {
                 self.sendLog(logBuffer.toString())
                 logBuffer.clear()
             }
         }
     }

  private func sendLog(_ message: String) {
    DispatchQueue.main.async {
      self.sendEvent(withName: "onNativeLog", body: ["log": message])
    }
  }
  
  override func supportedEvents() -> [String]! {
    return ["onNativeLog"]
  }
}


class StringBuffer {
    private var buffer: String = ""

    func append(_ string: String) {
        buffer += string
    }

    func toString() -> String {
        return buffer
    }

    func isNotEmpty() -> Bool {
        return !buffer.isEmpty
    }

    func clear() {
        buffer = ""
    }
}

File: android/app/src/main/java/com/nativecodedemo/NativeLogsModule.kt
package com.nativecodedemo

import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.io.BufferedReader
import java.io.InputStreamReader

class NativeLogsModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    @Volatile
    private var logging = false // Volatile variable to ensure visibility across threads
    private var logThread: Thread? = null // Thread for running the logging process

    override fun getName() = "NativeLogsModule" // Name of the module for React Native

    @ReactMethod
    fun startLogging() {
        if (logging) return // If already logging, do nothing
        logging = true // Set logging to true

        Log.d("NativeLogsModule", "Logging started") // Log to Android logcat
        sendLog("Logging started") // Send log to JavaScript

        logThread = Thread {
            try {
                // Execute logcat command to get system logs with timestamp
                val process = Runtime.getRuntime().exec("logcat -v time com.nativecodedemo:*")
                val bufferedReader = BufferedReader(InputStreamReader(process.inputStream))
                var line: String? = null // Initialize line variable
                val logBuffer = StringBuffer() // Buffer to collect logs
                while (logging && bufferedReader.readLine().also { line = it } != null) {
                    line?.let {
                        // Filter out logs generated by this module to prevent infinite logging loop
                        if (!it.contains("NativeLogsModule")) {
                            logBuffer.append(it).append("\n") // Append log to buffer
                        }
                    }
                    // Sleep for a second before sending logs
                    Thread.sleep(1000)
                    if (logBuffer.isNotEmpty()) {
                        sendLog(logBuffer.toString()) // Send buffered logs to JavaScript
                        logBuffer.setLength(0) // Clear the buffer
                    }
                }
                bufferedReader.close() // Close the bufferedReader
            } catch (e: Exception) {
                Log.e("NativeLogsModule", "Logging interrupted", e) // Log error to Android logcat
            }
        }
        logThread?.start() // Start the logging thread
    }

    @ReactMethod
    fun stopLogging() {
        if (!logging) return // If not logging, do nothing
        logging = false // Set logging to false
        logThread?.interrupt() // Interrupt the logging thread
        logThread = null // Set the thread to null

        Log.d("NativeLogsModule", "Logging stopped") // Log to Android logcat
        sendLog("Logging stopped") // Send log to JavaScript
    }

    private fun sendLog(message: String) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onNativeLog", message) // Emit event to JavaScript
    }
}
